derp/derp_server.go

```diff
@@ -1942,6 +2101,111 @@ func (s *Server) ServeDebugTraffic(w http.ResponseWriter, r *http.Request) {
+		for dstKey := range s.flows[c] {
+			snap[c][dstKey] = make(map[key.NodePublic]flowStats)
+			for srcOrFwd := range s.flows[c][dstKey] {
+				snap[c][dstKey][srcOrFwd] = *s.flows[c][dstKey][srcOrFwd]
+			}
+		}
+	}
+	return snap
+}
+
+// ServeFlows prints a list of recently active data flows through
+// this server, including the bandwidth used by each over the last
+// second.
+func (s *Server) ServeFlows(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "current time: %s\n", time.Now().Format("2006-01-02 15:04:05"))
+
+	lastTime := time.Now()
+	lastSnap := s.takeFlowsSnapshot()
+	time.Sleep(1 * time.Second)
+	currTime := time.Now()
+	currSnap := s.takeFlowsSnapshot()
+	dur := currTime.Sub(lastTime)
+
+	type flowLine struct {
+		connNum         int64
+		src, dst        key.NodePublic
+		fwdTo, recvFrom key.NodePublic
+		bytes           int64
+		packets         int64
+		bps             int64
+		pps             int64
+		lastActive      time.Duration
+	}
+
+	// Copy current flow snapshot into a sortable slice of flowLines
+	lines := make([]flowLine, 0)
+	for c := range currSnap {
+		for dstKey := range currSnap[c] {
+			for srcOrFwd := range currSnap[c][dstKey] {
+				f := currSnap[c][dstKey][srcOrFwd]
+				l := flowLine{
+					connNum:    c.connNum,
+					dst:        dstKey,
+					bytes:      f.bytes,
+					packets:    f.packets,
+					lastActive: currTime.Sub(f.lastActive),
+				}
+				// Calculate bandwidth
+				last := flowStats{}
+				if _, ok := lastSnap[c]; ok {
+					if _, ok2 := lastSnap[c][dstKey]; ok2 {
+						last = lastSnap[c][dstKey][srcOrFwd]
+					}
+				}
+				bytes := l.bytes - last.bytes
+				packets := l.packets - last.packets
+				l.bps = int64(float64(bytes) / dur.Seconds())
+				l.pps = int64(float64(packets) / dur.Seconds())
+				// src and mesh take a lil calculation
+				if srcOrFwd == (key.NodePublic{}) {
+					// common case, src and dst local
+					l.src = c.key
+				} else {
+					// one of src or dst is remote
+					if c.canMesh {
+						// dst is local, src is remote
+						l.recvFrom = c.key
+						l.src = srcOrFwd
+					} else {
+						// src is local, dst is remote
+						l.fwdTo = srcOrFwd
+						l.src = c.key
+					}
+				}
+				lines = append(lines, l)
+			}
+		}
+	}
+
+	fmt.Fprintf(w, "number flows: %v\n", len(lines))
+	// Sort by highest bandwidth flow
+	sort.Slice(lines, func(i, j int) bool { return lines[i].bps > lines[j].bps })
+
+	headStr := "%-10v %-10v %-10v %-10v %-10v %10v %10v %10v %10v %-12v\n"
+	bodyStr := "%-10v %-10v %-10v %-10v %-10v %10v %10v %10v %10v %5v secs\n"
+	fmt.Fprintf(w, headStr, "conn#", "src", "dst", "recv from", "fwded to", "bytes", "packets", "bps", "pps", "last active")
+	for _, l := range lines {
+		fmt.Fprintf(w, bodyStr, l.connNum, l.src.ShortString(), l.dst.ShortString(), l.recvFrom.ShortString(), l.fwdTo.ShortString(), l.bytes, l.packets, l.bps, l.pps, int(l.lastActive.Seconds()))
+	}
+
+	w.(http.Flusher).Flush()
```

Comment:
Is this flush needed? Normally the http library will flush after a synchronous http handler returns. If it is it would be great to have a comment, as this function otherwise looks like a normal synchronous handler.

Reply:
You are correct, it doesn't need a flush.